src/responder.ts

```diff
@@ -51,182 +61,357 @@ enum ResponderEvent {
+ */
+export class DoublingGasPolicy implements GasPolicy {
+    constructor(private readonly provider: ethers.providers.Provider) { }
+
+    getInitialPrice(): Promise<ethers.utils.BigNumber> {
+        return this.provider.getGasPrice();
+    }
+
+    getIncreasedGasPrice(previousPrice: ethers.utils.BigNumber): ethers.utils.BigNumber {
+        return previousPrice.mul(2);
+    }
+}
+
+/**
+ * A simple custom Error class to signal that the speified number of blocks has been mined.
+ */
+export class StuckTransactionError extends Error {
+   constructor(message: string) {
+       super(message);
+       this.name = "StuckTransactionError";
+   }
+}
+
+
+/**
+ * This class encapsulates the logic of trying to send a transaction and make sure it is mined with enough confirmations.
+ */
+export class EthereumTransactionMiner {
+
+    /**
+     * @param signer The Signer to use to send the transaction.
+     * @param confirmationsRequired The number of confirmations required.
+     * @param blocksThresholdForStuckTransaction The number of new blocks without the transaction is mined before considering
+     *                                           the transaction "stuck".
+     * @param newBlockTimeout The number of milliseconds since after which the provider is considered non-responsive.
+     */
+    constructor(
+        public readonly signer: ethers.Signer,
+        public readonly confirmationsRequired: number,
+        public readonly blocksThresholdForStuckTransaction: number,
+        public readonly newBlockTimeout: number
+    ) {}
+
+    /**
+     * This method sends the transaction, and resolves to the transaction hash.
+     * If an error is thrown, see ethers.js documentation for `JsonRpcSigner.sendTransaction`.
+     * @param transactionRequest The TransactionRequest to be sent.
+     */
+    public async sendTransaction(transactionRequest: ethers.providers.TransactionRequest): Promise<string> {
+        const txResponse = await this.signer.sendTransaction(transactionRequest);
+        return txResponse.hash;
+    }
+
+    /**
+     * Resolves after the transaction receives the first confirmation.
+     * Rejects with `NoNewBlockError` the provider does not receive a new block for `newBlockTimeout` milliseconds.
+     * Rejects with `BlockThresholdReachedError` if the transaction is still unconfirmed
+     * after `blocksThresholdForStuckTransaction` blocks are mined.
+     *
+     * @param txHash The transaction hash, returned by `sendTransaction`.
+     * @param timeLastBlockReceived Optional; if known, the time when the last block was received by the provider.
+     *                              The provider will be considered unresponsive after `newBlockTimeout` from `timeLastBlockReceived`.
+     *                              If not given, the current time returned by `Date.now()` will be used.
+     */
+    public async waitForFirstConfirmation(txHash: string, timeLastBlockReceived: number = Date.now()) {
```

Comment:
This default isnt used

Reply:
I thought it makes sense here to use the current time as a sensible default value, in case you are using this function in a situation where you don't know/care when was the last block. If you disagree, I can just remove it for now.

Reply:
Yeah, I can see why it might become a useful default. But I would wait until someone needs it, and let them add it. I actually kinda hate defaults as I've seen them be the source of a few annoying bugs, so I only kinda use them if necessary.

Reply:
Removed in 352fc270d00d7838ef20e737c027295a3bb1c97d.